### ISSUE TITLE
Implement service-apis behavior

### ISSUE THREAD

**Issue by youngnick** [projectcontour/contour#2287]
Once the behavior of the service-apis objects is decided, this issue will cover actually implementing that behavior.

**Comment by youngnick** [comment]
Blocked on service-apis actually defining the behaviour.

**Comment by danehans** [comment]
[v1alpha1](https://docs.google.com/document/d/1BJoxZTbh2bkYQnqVOETm6_7q4hjwGop2pTZDc__5olI/edit) of service apis is scheduled to be released this month (August 2020). As discussed during this week's community call, it would be helpful to begin planning for the implementation. @jpeach suggested that work such as https://github.com/projectcontour/contour/issues/2683 be completed prior to the service apis implementation. Should this issue be used as the meta issue for describing a) the work that needs to be completed prior to implementing service apis and b) the actual service api's implementation? Should a separate issue be created for a? In general, I would like to enumerate the work needed to implement service-apis, including anything that's needed prior to the actual implementation so myself and others can begin helping with the service apis implementation.

/cc @Miciah @stevesloka

**Comment by youngnick** [comment]
Thanks @danehans, I agree that this issue should hold:
- a list of issues that block doing some implementation of the service apis behavior
- a list of issues that need to be implemented to actually deliver the behavior as at v1alpha1.

We should figure those out here, then I will add them to the initial description as a checklist so we get a count.

**Comment by jpeach** [comment]
Blocked on https://github.com/projectcontour/contour/issues/2226, which will allow new builders to be added at runtime.

**Comment by youngnick** [comment]
- [ ] Decision about *Class behavior (IngressClass and GatewayClass)
- [ ] Work to add builder interface (@stevesloka)
- [ ] Work to add interface between DAG and xDS caches (@jpeach)

**Comment by youngnick** [comment]
Okay, for the service-apis objects to be supported in Contour, we will need:
- [ ] GatewayClass watching support. Note that as decided in #2809, only one GatewayClass is supported per Contour deployment. The best way to make this work is TBD.
- [ ] Gateway watching support. Note that Contour itself will not create Gateways, it will watch for Gateways in its GatewayClass, and build a configuration to send valid Listeners to Envoy, writing back information about Listener status to the relevant Gateway.
- [ ] HTTPRoute watching support. Contour should take HTTPRoutes referenced from Listeners, and add config to Envoy to expose the routes.
- [ ] TLSRoute watching support. Contour should take TLSRoutes referenced from Listeners, and add config to Envoy to expose the routes.

These are the only resources that Contour needs to implement for its service-apis support. Note that we will probably need to work closely with upstream to define what conformance means for this feature set (that is, the "ingress controller" feature set, or some similar name).

**Comment by shashankram** [comment]
I am interested in exploring the use of Contor for ingress using the Gateway API. Is there a document that describes how much of the API is supported, and what the roadmap is? It seems like there is functionality that is usable based on https://projectcontour.io/guides/gateway-api/, just wasn't sure exactly what is supported at the moment.

**Comment by youngnick** [comment]
We currently don't have a canonical document aside from this issue, as the support is changing too quickly. We do only plan to support HTTPRoute and TLSRoute as route targets however, in line with our positioning of Contour as a layer 7 ingress controller.

**Comment by skriss** [comment]
Is anyone opposed to closing this tracking issue now that our support is pretty far along, and managing the remaining backlog solely via the [gateway-api label](https://github.com/projectcontour/contour/labels/area%2Fgateway-api)?
